core/src/replay_stage.rs

```diff
@@ -1172,12 +1175,292 @@ pub(crate) mod tests {
         system_transaction,
         transaction::TransactionError,
     };
-    use solana_vote_program::vote_state::VoteState;
+    use solana_stake_program::stake_state;
+    use solana_vote_program::vote_state;
+    use solana_vote_program::vote_state::{Vote, VoteState};
+    use std::iter;
     use std::{
         fs::remove_dir_all,
         sync::{Arc, RwLock},
     };
 
+    struct ForkInfo {
+        leader: usize,
+        fork: Vec<Slot>,
+        voters: Vec<usize>,
+    }
+
+    struct ValidatorInfo {
+        stake: u64,
+        keypair: Keypair,
+        voting_keypair: Keypair,
+        staking_keypair: Keypair,
+    }
+
+    struct ForkSelectionResponse {
+        slot: u64,
+        is_locked_out: bool,
+    }
+
+    fn simulate_fork_selection(
```

Comment:
sorry to jump in at the last part of the review, but typically we want to avoid these huge test harnesses.  It should be a signal that the code you are testing is poorly written and the interface is not setup for easy unit testing.  We should refactor `select_fork` to be more opaque and do its work over a simpler data structure that doesn't deal with vote states, banks etc....

Reply:
@aeyakovenko makes sense. Let's merge this now, for the time being. And in next PR, I can refactor the function (with this attack protection) and update test harness.

Reply:
Sounds good.   The “right way” to do this is to split out “ForkSelector” as a separate type and parametrize replay stage with it.  Unit test the module and build an “AttackingForkSelector” that does the desired wrong behavior.  Then make it possible to configure the validator in LocalCluster with either.